### PR TITLE
Add UnderscorePluralNamingStrategy

### PR DESCRIPTION
**Pull request by zorn-v**
Converts 'MyEntity' to 'my_entities' or 'MY_ENTITIES'.

### CODE REVIEW

**Review by SenseException — CHANGES_REQUESTED** [review]
Thank you for your contribution. I'm not sure if this feature will be added to the core code of ORM. It is basically the `UnderscoreNamingStrategy` code with a pluralisation.

I would expect it to allow other strategies to be used for pluralisation too instead for only the underscore one. This PR is also missing tests.

Further reviews are required before the new strategy gets accepted by other reviewers.

**Comment by zorn-v** [comment]
Helooooo. No missing tests about month

**Comment by SenseException** [comment]
> I would expect it to allow other strategies to be used for pluralisation too instead for only the underscore one.

Thanks for adding the tests. As mentioned before this class is a copy of the UnderscoreNamingStrategy and therefore duplicated code. Can this pluralisation also work for other naming strategies?

**Comment by zorn-v** [comment]
> Can this pluralisation also work for other naming strategies?

You can ask it of architect designer of doctrine. I think no. You must create another *NamingStrategy.
But who names tables with no "underscore plural" now ? 
Yep, philosophical question :smile:

**Comment by zorn-v** [comment]
Main purpose of this PR is to make possible configurable `things` instead of `thing` (table has MANY entities) default table names in symfony doctrine bundle. And do not provide `@ORM\Table(name="things")` in EACH entity

**Comment by greg0ire** [comment]
> You can ask it of architect designer of doctrine ORM. I think no. You must create another *NamingStrategy.

I think it can and should be done, by injecting an optional dependency in the existing strategies and delegating the `classToTableName` conversion responsability to it. Does that make me an architect designer?

**Comment by zorn-v** [comment]
Yep, my fail. In doctrine bundle we can make service like "doctrine.orm.naming_strategy.underscore_plural" with same class (%doctrine.orm.naming_strategy.underscore.class%) and constructor parameters then

**Comment by zorn-v** [comment]
Add plural constructor parameter for naming strategies

File: lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php [hunk 1]
```diff
@@ -100,18 +121,37 @@ public function joinColumnName($propertyName, $className = null)
...
+        if (strpos($className, '\\') !== false) {
+            $className = substr($className, strrpos($className, '\\') + 1);
+        }
+
+        if ($pluralize) {
+            $className = Inflector::pluralize($className);
+        }
+
+        return $this->underscore($className);
+    }
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: Can we avoid this duplication of code?

Comment (zorn-v) [reply lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: You mean something like protected method in default strategy and `return $this->underscore(parent::_classToTableName($className, $pluralize));` in underscore strategy, and inherit underscore from default ?
Is it really needed ?

Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: I mean a new dependency you would inject in the constructors of the existing strategies, and initialize to a default implementation if not provided.

Comment (zorn-v) [reply lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: Well `$plural = false` as constructor argument - is not what you mean or what ?

Comment (zorn-v) [reply lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: Oh, seems I got your point - remove "UnderscoreNamingStrategy" and add constructor parameter for default one like "$underscore = false" ? But it breaks BC very... much ) I think

Comment (zorn-v) [reply lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: But... for BC we can make `parent::__construct` with proper parameters :smile: 
Do I understand you correctly ?

Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: Almost… I mean this:
```
    public function __construct(/*parentParams*/, ?ClassToTableNameTransformer $classToTableNameTransformer = null)
    {
        parent::__construct($parentParams);

        if ($classToTableNameTransformer) {
            $this->classToTableNameTransformer = $classToTableNameTransformer;

            return;
        }
        $this->classToTableNameTransformer = new DefaultClassToTableNameTransformer();
    }
```

Comment (zorn-v) [reply lib/Doctrine/ORM/Mapping/Factory/UnderscoreNamingStrategy.php]: One more class just for "class for one more class" ? Java bite you :smile:

**Comment by beberlei** [comment]
This could possibly be implemented as a decorator;

```php
$strategy = new PluralizeTableNameStrategy(new DefaultNamingStrategy);
```
The benefit would be the existing code wouldn't need to change at all (less risk of breakage). Another benefit then it wouldn't need the isPlural and setPlural methods, because that follows from using the class.

Also the tests added are without significane now, because even though plural=true none of them actually have a term that gets pluralized. Please use something meaningful like the MyEntity => my_entities example you gave in the PR description.
